### PR TITLE
whizard: add gosam variant

### PR DESCRIPTION
**Pull request by vvolkl**

### CODE REVIEW

**Comment by vvolkl** [comment]
@spackbot fix style

File: var/spack/repos/builtin/packages/whizard/package.py [hunk 1]
```diff
@@ -141,15 +146,19 @@ def configure_args(self):
...
+        args.extend(self.enable_or_disable("gosam"))
         args.extend(self.enable_or_disable("pythia8"))
         args.extend(self.enable_or_disable("lcio"))
         args.extend(self.enable_or_disable("lhapdf"))
         args.extend(self.enable_or_disable("openloops"))
 
+        if "+gosam" in spec:
+            args.append(f"--with-golem={spec['gosam-contrib'].prefix}")
+            args.append(f"--with-gosam={spec['py-gosam'].prefix}")
+            args.append(f"--with-qgraf={spec['qgraf'].prefix}")
```
Comment (tmadlener) [inline var/spack/repos/builtin/packages/whizard/package.py]: Are these actually used during configuration? Or is whizard also using environment variables instead now to find these?

File: var/spack/repos/builtin/packages/qgraf/package.py [hunk 2]
```diff
@@ -20,4 +20,6 @@ class Qgraf(Package):
     def install(self, spec, prefix):
         fortran = Executable(spack_fc)
         fortran("qgraf-{0}.f".format(self.spec.version), "-o", "qgraf")
-        install_tree(".", prefix)
+        # TODO: use share... subdirectories
```
Comment (tmadlener) [inline var/spack/repos/builtin/packages/qgraf/package.py]: I am not sure I understand the TODO here. Should this be installed into `<prefix>/share`?

Comment (vvolkl) [reply var/spack/repos/builtin/packages/qgraf/package.py]: Well, the executable is fine in bin, but the `install_tree` brings along the kitchen cabinet:
```
[root@fcc-spack-01 ~]# ls /root/spack/opt/spack/linux-almalinux9-cascadelake/gcc-11.4.1/qgraf-3.4.2-kpkqltneoyvbbuikdqpnhpn4tmurywrx/bin
array.sty  form.sty  phi3  qcd  qed  qgraf  qgraf-3.0.pdf  qgraf-3.4.2.f  qgraf-3.4.2.pdf  qgraf.dat  sum.sty
```
The manual .pdf could go to share for sure (or be omitted), for the other files I'm not sure if they are needed at runtime by qgraf, so I left them.

File: var/spack/repos/builtin/packages/whizard/package.py (hunk 1 above)
Comment (vvolkl) [reply var/spack/repos/builtin/packages/whizard/package.py]: Whizard can use both `$GOSAM_DIR` and `--with-gosam`, not sure of the order of precedence. I'd have to doublecheck if they are needed, I put those in when debugging the missing `golem.la` whizard was expecting.

**Comment by vvolkl** [comment]
Actually, this does not work yet - gosam-contrib puts the compiler wrapper path in `gosam.conf`, still need to patch that out.

File: var/spack/repos/builtin/packages/qgraf/package.py (hunk 2 above)
Comment (vvolkl) [reply var/spack/repos/builtin/packages/qgraf/package.py]: just installed the  qgraf exe now, whizard still works fine. If you want the manual, it's in the source tarball anyway.

File: var/spack/repos/builtin/packages/whizard/package.py (hunk 1 above)
Comment (vvolkl) [reply var/spack/repos/builtin/packages/whizard/package.py]: Turns out that the spack sets up the dependencies correctly, these lines weren't needed. Good catch!

**Comment by vvolkl** [comment]
Comments adressed, compiler wrapper path removed.

**Review by tmadlener — APPROVED** [review]

File: var/spack/repos/builtin/packages/gosam-contrib/package.py [hunk 3]
```diff
@@ -26,6 +29,11 @@ class GosamContrib(AutotoolsPackage):
     )
     variant("pic", default=False, description="Build position-independent code")
 
+    def patch(self):
+        # remove spack compiler wrapper path
+        mf = FileFilter("gosam.conf.in")
+        mf.filter("^fc.bin=.*", "fc.bin=gfortran")
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/gosam-contrib/package.py]: Can we use e.g. `self.compiler.fc` instead of hardcoding `gfortran`?

**Review by alalazo — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/gosam-contrib/package.py (hunk 3 above)
Comment (vvolkl) [reply var/spack/repos/builtin/packages/gosam-contrib/package.py]: Yes, good call. Took a while, but I tested it now.
```suggestion
        mf.filter("^fc.bin=.*", "fc.bin=" + self.compiler.fc)
```

**Review by alalazo — APPROVED** [review]

**Pull request merged**